### PR TITLE
Powchain service reconnect

### PR DESCRIPTION
**Pull request by shayzluf**
[Resolves] #2108

---

# Reconnect go-ethereum rpc client on EOF

When an rpc client disconnects (EOF) it doesn't reconnects this PR is to fix this behavior

### CODE REVIEW

File: beacon-chain/powchain/service.go [hunk 1]
```diff
@@ -329,8 +358,14 @@ func (w *Web3Service) run(done <-chan struct{}) {
 			log.Debug("ETH1.0 chain service context closed, exiting goroutine")
 			return
 		case w.runError = <-headSub.Err():
-			log.Debugf("Unsubscribed to head events, exiting goroutine: %v", w.runError)
-			return
+			if w.runError.Error() == "EOF" {
```
Comment (prestonvanloon) [inline beacon-chain/powchain/service.go]: Import io and compare against `io.EOF`

File: beacon-chain/powchain/service.go [hunk 2]
```diff
@@ -329,8 +358,14 @@ func (w *Web3Service) run(done <-chan struct{}) {
...
+				log.Infof("Got EOF: restarting rpc client")
```
Comment (prestonvanloon) [inline beacon-chain/powchain/service.go]: This should be a WARN, please use `log.Warn(...`

File: beacon-chain/powchain/service_test.go [hunk 3]
```diff
@@ -264,6 +264,51 @@ func TestStop_OK(t *testing.T) {
...
+	// time.Sleep(5 * time.Second)
+	// msg = hook.LastEntry().Message
+	// want = "EOF"
+	// if msg != want {
+	// 	t.Errorf("incorrect log, expected %s, got %s", want, msg)
+	// }
+	// newClientPTR := &web3Service.client
+	// if clientPTR == newClientPTR {
+	// 	t.Errorf("client wasn't reinitialized after EOF")
+	// }
```
Comment (prestonvanloon) [inline beacon-chain/powchain/service_test.go]: What's up with this commented code?

File: beacon-chain/powchain/service_test.go [hunk 4]
```diff
@@ -338,6 +383,7 @@ func TestStatus(t *testing.T) {
 		{isRunning: true, blockTime: beforeFiveMinutesAgo}: "eth1 client is not syncing",
 		{isRunning: true}: "eth1 client is not syncing",
 		{isRunning: true, runError: errors.New("test runError")}: "test runError",
+		{isRunning: true, runError: errors.New("EOF")}:           "EOF",
```
Comment (prestonvanloon) [inline beacon-chain/powchain/service_test.go]: io.EOF

File: beacon-chain/powchain/service.go [hunk 5]
```diff
@@ -284,6 +288,32 @@ func (w *Web3Service) handleDelayTicker() {
 		log.Error(err)
 	}
 }
+func (w *Web3Service) rerunRPCClientAndSubscribe(headSub ethereum.Subscription, header *gethTypes.Header) {
+	rpcClient, err := gethRPC.Dial(w.endpoint)
+	if err != nil {
+		log.Errorf("Access to PoW chain is required for validator. Unable to connect to Geth node: %v", err)
```
Comment (prestonvanloon) [inline beacon-chain/powchain/service.go]: Shouldn't you set the `w.runError` and return?

Comment (prestonvanloon) [inline beacon-chain/powchain/service.go]:
```suggestion
		log.Errorf("Access to PoW chain is required for validator. Unable to connect to ETH 1 node: %v", err)
```

File: beacon-chain/powchain/service.go [hunk 6]
```diff
@@ -284,6 +288,32 @@ func (w *Web3Service) handleDelayTicker() {
...
+	}
+	header, err = w.blockFetcher.HeaderByNumber(w.ctx, nil)
+	if err != nil {
+		log.Errorf("Unable to retrieve latest ETH1.0 chain header: %v", err)
+		w.runError = err
+		return
+	}
+	w.blockHeight = header.Number
+	w.blockHash = header.Hash()
+}
```
Comment (prestonvanloon) [inline beacon-chain/powchain/service.go]: Clear out `w.runError` if you are able to successfully resubscribe. Having a w.runError is an unhealthy condition and the process may be terminated if it cannot recover in a reasonable amount of time.

File: beacon-chain/powchain/service_test.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/powchain/service_test.go]: couldn't make it send eof error to the header subscription

File: beacon-chain/powchain/service.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/powchain/service.go]: thanks. done

File: beacon-chain/powchain/service.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/powchain/service.go]: thanks

File: beacon-chain/powchain/service.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/powchain/service.go]: yes

File: beacon-chain/powchain/service.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/powchain/service.go]: done

File: beacon-chain/powchain/service_test.go (hunk 3 above)
Comment (prestonvanloon) [reply beacon-chain/powchain/service_test.go]: you can't do `gr.feed.Send(io.EOF)`?

**Pull request closed without merging**